blockchain/blockchain.go

```diff
@@ -765,15 +770,47 @@ func revertHead(txn db.Transaction) error {
 	if err = txn.Delete(db.BlockHeaderNumbersByHash.Key(header.Hash.Marshal())); err != nil {
 		return err
 	}
+	if blockNumber != 0 { // Don't send block before genesis
+		var newHeader *core.Header
+		newHeader, err = blockHeaderByNumber(txn, blockNumber-1)
+		if err != nil {
+			return err
+		}
+		b.newHeads.Send(newHeader)
+	}
+
+	if err = removeTxsAndReceipts(txn, blockNumber, header.TransactionCount); err != nil {
+		return err
+	}
+
+	// remove state update
+	if err = txn.Delete(db.StateUpdatesByBlockNumber.Key(numBytes)); err != nil {
+		return err
+	}
+
+	// remove pending
+	if err = txn.Delete(db.Pending.Key()); err != nil {
+		return err
+	}
+
+	// update chain height
+	if blockNumber == 0 {
+		return txn.Delete(db.ChainHeight.Key())
+	}
 
+	heightBin := core.MarshalBlockNumber(blockNumber - 1)
+	return txn.Set(db.ChainHeight.Key(), heightBin)
+}
+
+func removeTxsAndReceipts(txn db.Transaction, blockNumber, numTxs uint64) error {
 	blockIDAndIndex := txAndReceiptDBKey{
 		Number: blockNumber,
 	}
 	// remove txs and receipts
-	for i := uint64(0); i < header.TransactionCount; i++ {
+	for i := uint64(0); i < numTxs; i++ {
 		blockIDAndIndex.Index = i
 		var reorgedTxn core.Transaction
```

Comment:
we don't need this declaration because below is `:=` assignment now

Reply:
Sorry, I don't understand.

Reply:
this line is not used, before your changes it was necessary because `=` is assignment operator, but `:=` assignment + declaration operator (terminology is mine, but I think you get idea).

Reply:
I see now. I didn't realize you were referring to the `var reorgedTxn core.Transaction` line. Fixed now, thanks.